### ISSUE TITLE
Enable preview fields for keys?

### ISSUE THREAD

**Issue by chrisjbremner** [cloudhut/kowl#219]
How would you feel about enabling a preview field for keys as well as values? It's possible to have a map as a key, so it might be helpful to filter out some unneeded info there too.

**Comment by rikimaru0345** [comment]
You mean the key of a kafka message, right?

Hm, why not, I don't have anything against that. I think it it would be no problem to also show the preview tags in the key column.

But while the feature itself is probably super simple to do, it opens a few UI/UX issues:
Currently the button to configure the preview fields is in the title of the value column. Should that button be moved somewhere else? If so, where??

**Comment by chrisjbremner** [comment]
In my head I was envisioning something like this which seems like the simplest to implement. Otherwise, I suppose it could live next to the topic name or next to the filter switch (although this would require you to add some dropdown to assign each preview to either key/value)

![preview_keys](https://user-images.githubusercontent.com/15189285/118522669-2d668880-b6f1-11eb-983e-f09f69bab66a.png)

**Comment by rikimaru0345** [comment]
Next version will have more options for where to search.
I didn't include searching in headers for now, but that can be added if there is need.

![image](https://user-images.githubusercontent.com/28133019/119357209-95940c00-bca7-11eb-9976-e1936db9d81a.png)

**Comment by chrisjbremner** [comment]
Looks great, thanks! One small feedback is that it appears to shrink the column width when using the preview, such that you can't see the preview value.
![image](https://user-images.githubusercontent.com/15189285/119367841-3adac000-bc67-11eb-88c1-85a78fb9da07.png)

**Comment by rikimaru0345** [comment]
Ok, I'll see what I can do about that.

**Comment by rikimaru0345** [comment]
Html tables don't have any way to distribute width as intended.
They can't really truncate text based on the content as far as I know.
I just pushed a commit that either distributes the width either "50:50" when both columns have preview tags active, or "25:auto" when not.

Can you check if that works for you?
We might have to rewrite all tables to use display:grid in the future to fix issues like that.

**Comment by chrisjbremner** [comment]
That works for me for now, although it'd be nice if the key column would do auto-width like it does without any preview active.

Here's what the behavior was before without a key preview:
![before_nopreview](https://user-images.githubusercontent.com/15189285/119399289-915af500-bc8d-11eb-9f02-b52445d2a30b.png)

Here's what it was in `master-82ab659a` (before your last change) with a key preview (content gets cut off):
![before](https://user-images.githubusercontent.com/15189285/119399268-8b651400-bc8d-11eb-9641-39be06be4688.png)

Here's what it is with the latest commit and a key preview (lots of extra key whitespace, our values generally have more content so would prefer that to consume the 'leftovers' from all of the other columns):
![after](https://user-images.githubusercontent.com/15189285/119399373-aa63a600-bc8d-11eb-9067-a7a6101a1acf.png)

**Comment by rikimaru0345** [comment]
Sorry, at least right now I can't implement a better way than that.
There is no way to easily configure the columns widths so each column has an "auto" width (using only as much space as it needs), while also guaranteeing that the columns won't grow out of the table.

Using `px` (or any other absolute size) is obviously bad since it doesn't react to anything. The width of the window, table, and content are not considered at all.

Using `%`  is almost as bad. While it does distribute the column widths according to the given percentages, it doesn't consider the content, leading to potentially wasted space.

Using `auto` only works fine when the content of the columns is small. But when the content of a cell is large, the column will just grow without any limits, happily taking up more space than the table - even if the column and/or the table have a max-width defined.

The biggest issue is the table columns growing beyond the table, that's why we're currently using a hack to at least get the cells to properly make use of wrapping and truncation.


I've known about this problem for quite some time and I have some avenues to explore on my todo list.
Maybe I'll create a completely custom table component that - instead of the standard html - uses `display:grid` instead, which has a lot more options to control column widths.
Or maybe I'll implement a way to make columns resizable directly (by clicking/dragging).
Another alternative would be to just let the table expand and add proper scrollbars, however that makes for a pretty bad user experience and will also break wrapping/truncation completely...

So for now, tweaking the percentage ratios is unfortunately the only thing I can offer. 
It sucks, I know, but I'll definitely work on fixing it in the future (which most likely won't be too far off).

**Comment by chrisjbremner** [comment]
Makes sense, thanks for taking a look at it! I'll close this ticket since the original request was fulfilled
